pkg/detection/detect.go

```diff
@@ -0,0 +1,158 @@
+package detection
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+)
+
+type FileType string
+
+const (
+	FileTypeCloudFormation FileType = "cloudformation"
+	FileTypeTerraform      FileType = "terraform"
+	FileTypeDockerfile     FileType = "dockerfile"
+	FileTypeKubernetes     FileType = "kubernetes"
+	FileTypeYAML           FileType = "yaml"
+	FileTypeTOML           FileType = "toml"
+	FileTypeJSON           FileType = "json"
+)
+
+var matchers = map[FileType]func(name string, r io.Reader) bool{}
+
+func init() {
+
+	matchers[FileTypeJSON] = func(name string, r io.Reader) bool {
+		ext := filepath.Ext(filepath.Base(name))
+		return strings.EqualFold(ext, ".json")
+	}
+
+	matchers[FileTypeYAML] = func(name string, r io.Reader) bool {
+		ext := filepath.Ext(filepath.Base(name))
+		return strings.EqualFold(ext, ".yaml") || strings.EqualFold(ext, ".yml")
+	}
+
+	matchers[FileTypeTOML] = func(name string, r io.Reader) bool {
+		ext := filepath.Ext(filepath.Base(name))
+		return strings.EqualFold(ext, ".toml")
+	}
+
+	matchers[FileTypeTerraform] = func(name string, _ io.Reader) bool {
+		ext := filepath.Ext(filepath.Base(name))
+		return strings.EqualFold(ext, ".tf") || strings.EqualFold(ext, ".tf.json")
+	}
+
+	matchers[FileTypeCloudFormation] = func(name string, r io.Reader) bool {
+		var unmarshalFunc func([]byte, interface{}) error
+
+		switch true {
+		case IsType(name, r, FileTypeYAML):
+			unmarshalFunc = yaml.Unmarshal
+		case IsType(name, r, FileTypeJSON):
+			unmarshalFunc = json.Unmarshal
+		default:
+			return false
+		}
+
+		if r == nil {
+			return false
+		}
+
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return false
+		}
+
+		contents := make(map[string]interface{})
+		if err := unmarshalFunc(data, &contents); err != nil {
+			return false
+		}
+		_, ok := contents["Resources"]
+		return ok
+	}
+
+	matchers[FileTypeDockerfile] = func(name string, _ io.Reader) bool {
+		const requiredFile = "Dockerfile"
+		base := filepath.Base(name)
+		ext := filepath.Ext(base)
+		if strings.TrimSuffix(base, ext) == requiredFile {
```

Comment:
I wonder if this should just be a strings.HasPrefix(name, "Dockerfile") because I think we use `Dockerfile.CI` 

Not sure if there is a convention for this naming though

Reply:
I think that will do it - it will trim the extension off (`.CI`) and then `Dockerfile == Dockerfile`. I did it like that so we don;t match stuff like `dockerfile.go`.